### PR TITLE
[DOP-20061] Stats for relations response

### PR DESCRIPTION
**Pull request by TiGrib**
## Change Summary

Add stats for `Output` and `Input` relations.
`num_bytes`, `num_files`, `num_rows`.

## Related issue number

[DOP-20061]

## Checklist

* [ ] Commit message and PR title is comprehensive
* [ ] Keep the change as small as possible
* [ ] Unit and integration tests for the changes exist
* [ ] Tests pass on CI and coverage does not decrease
* [ ] Documentation reflects the changes where applicable
* [ ] `docs/changelog/next_release/<pull request or issue id>.<change type>.rst` file added describing change
  (see [CONTRIBUTING.rst](https://github.com/MobileTeleSystems/data-rentgen/blob/develop/CONTRIBUTING.rst) for details.)
* [ ] My PR is ready to review.

### CODE REVIEW

File: data_rentgen/server/schemas/v1/lineage.py [hunk 1]
```diff
@@ -116,12 +106,39 @@ class RunLineageQueryV1(BaseLineageQueryV1):
 
 
 class LineageRelationV1(BaseModel):
```
Comment (dolfinus) [inline data_rentgen/server/schemas/v1/lineage.py:108]: Actually, this class could be removed. Just copy `to` and `from_` fields to inherited models

File: data_rentgen/server/schemas/v1/lineage.py [hunk 2]
```diff
@@ -116,12 +106,39 @@ class RunLineageQueryV1(BaseLineageQueryV1):
...
+class LineageSymlinkRelationV1(LineageRelationV1):
+    kind: Literal["SYMLINK"] = "SYMLINK"
+    type: str = Field(description="Type of relation between datasets", examples=["METASTORE", "WAREHOUSE"])
 
 
 class LineageResponseV1(BaseModel):
-    relations: list[LineageRelationV1] = []
+    relations: list[
+        LineageParrentRelationV1 | LineageInputRelationV1 | LineageOutputRelationV1 | LineageSymlinkRelationV1
+    ] = []
```
Comment (dolfinus) [inline data_rentgen/server/schemas/v1/lineage.py:143]: nit: not related to this PR, but I recommend replacing `= []` with `= Field(default_factory=list)`

File: tests/test_server/utils/stats.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
+from data_rentgen.db.models import Input, Output
+
+
+async def relation_stats(relations: list[Input] | list[Output]):
```
Comment (dolfinus) [inline tests/test_server/utils/stats.py:4]: There is no IO or network calls in this method, so it should be sync

File: tests/test_server/utils/stats.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        stats[key]["num_files"] += relation.num_files
+        stats[key]["created_at"] = max(stats[key]["created_at"] or relation.created_at, relation.created_at)
+
+    return stats
+
+
+async def relation_stats_by_runs(relations: list[Input] | list[Output]):
+    stats = {}
+    for relation in relations:
+        key = (str(relation.run_id), relation.dataset_id)
```
Comment (dolfinus) [inline tests/test_server/utils/stats.py:39]: Why converting keys to string?

Comment (TiGrib) [reply tests/test_server/utils/stats.py:39]: I make them string for simultanious. I use them in assertion's where we cast run_id and operation_id to strings.
```python
           {
                "kind": "INPUT",
                "from": {"kind": "DATASET", "id": dataset.id},
                "to": {"kind": "OPERATION", "id": str(operation.id)},
                "num_bytes": input_stats[(str(operation.id), dataset.id)]["num_bytes"],
                "num_rows": input_stats[(str(operation.id), dataset.id)]["num_rows"],
                "num_files": input_stats[(str(operation.id), dataset.id)]["num_files"],
                "last_interaction_at": input_stats[(str(operation.id), dataset.id)]["created_at"].strftime(
                    "%Y-%m-%dT%H:%M:%S.%fZ",
                ),
            }
    ```

Comment (dolfinus) [reply tests/test_server/utils/stats.py:39]: Why not using `input_stats[(operation.id, dataset.id)]`? Why wrapping operation.id with str()?

File: tests/test_server/test_lineage/test_job_lineage.py [hunk 5]
```diff
@@ -232,7 +238,7 @@ async def test_get_job_lineage(
             }
             for dataset in sorted(datasets, key=lambda x: x.id)
         ],
-    }
+    }, output_stats
```
Comment (dolfinus) [inline tests/test_server/test_lineage/test_job_lineage.py:241]:
```suggestion
    }
```

**Review by dolfinus — APPROVED** [review]

**Pull request merged**